src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -554,31 +562,39 @@ class GroupReadsByUmi
   }
 
   /**
-    * Retrieves the UMI to use for a read pair.  In the case of single-umi strategies this is just
+    * Retrieves the UMI to use for a template.  In the case of single-umi strategies this is just
     * the upper-case UMI sequence.
     *
     * For Paired the UMI is extended to encode which "side" of the template the read came from - the earlier
     * or later read on the genome.  This is necessary to ensure that, when the two paired UMIs are the same
     * or highly similar, that the A vs. B groups are constructed correctly.
     */
-  private def umiForRead(r1: SamRecord, r2: SamRecord): Umi = r1.get[String](this.rawTag) match {
-    case None | Some("") => fail(s"Record '$r1' was missing the raw UMI tag '${this.rawTag}'")
-    case Some(chs)       =>
-      val umi = chs.toUpperCase
+  private def umiForRead(t: Template): Umi = {
+    // Check that all the primary reads have the UMI defined
+    t.primaryReads.foreach { rec =>
+      val umi = rec.get[String](this.rawTag)
+      if (!umi.exists(_.nonEmpty)) fail(s"Record '$rec' was missing the raw UMI tag '${this.rawTag}'")
```

Comment:
Why not `require`?  When is something "non-exceptional"?

Reply:
`require` and `fail` both throw exceptions.  It's just that in FgBioMain we caatch `FailureException` and print a more nicely formatted message without the stack trace.  So I think `fail` is nicer if the error is likely due to user input.